### DISCUSSION TITLE
Building a modified Steam Plugin for testing/release?

### DISCUSSION THREAD

**Discussion by ABaumher** [GOG-Nebula/.github#14] [Q&A] [answered]
This is something of a branch off of #2 but i didn't want to bump an old thread. I've been working on fixing the Steam Integration, and i'm at the point where i've gotten over the initial hurdle of figuring out what is actually going on in the plugin and why it's coded as it is (frankly, i still don't know, but i can work with it so it's irrelevant). I am testing the initial calls now, and should the first one work, getting the rest in play will be a much quicker process. It's not guarenteed; i may hit some unseen snag or the code could just not work, but I'm at the point where if i want to let others test it for me, i need to worry about deployment. The code has evolved to a point where 'comment out/replace this line' isn't going to work, as it's like 20 files and we also need an additional package (RSA), which requires including the python module directly in the plugin. I've been doing this manually for testing, but I can't realistically tell other people to aid in testing or work on other parts, let alone when we release. 

Is there a tasks.py/generic buildtools script i can use as a basis for the steam plugin? Please note i haven't seen any documentation for buildtools and have no idea how it works. I probably could decode it if given enough time, but i really don't want to. I have also never done anything with github workflows, so that code from UPlay is gibberish to me. Again, i could probably spend a while googling how it works (and there may be a tutorial, idk), but i'd very much like to focus on getting the plugin working because it's not a simple fix. 

Edit: spelling errors

**Comment by ABaumher** [comment]
An aside: if anyone here wants to help, we have a discussion where we explain what we're doing and testing at https://github.com/FriendsOfGalaxy/galaxy-integration-steam/discussions/164. 
The rough discovery of what was busted is discussed at https://github.com/FriendsOfGalaxy/galaxy-integration-steam/issues/159 
The main repo where we post stuff is at https://github.com/slo/galaxy-integration-steam

Another aside: I'd potentially need to do this for the PSN Plugin as well. From what i can tell, the Akamai authentication Sony uses flags the web browser GOG Galaxy uses, so logging in hangs with a security error. This is pretty common from what i can tell, the workaround most projects use is to manually have the user copy an NPSSO cookie after logging in via chrome/edge/etc. I've attempted to circumvent this in various ways, notably using a tab to communicate back to the plugin and even running a local web server in hopes of getting the cookie via an exposed request, but their security is actually pretty good (even if i bypassed CORS policy in the browser, their server would detect the CORS mismatch). I could change the PSN plugin to simply have an input field and have them do what everyone else does and paste that token, if that's a good idea. But, if i did, i'd need a build tool for that plugin as well

**Comment by urwrstkn8mare** [comment, marked as answer]
@ABaumher try using the `tasks.py` (and check any other changes made for it to work) from here https://github.com/thibmo/galaxy-integration-steam/blob/fix_steam_2023_01. The `README.md` lays out how to use it (`inv`). The instructions say to create a virtualenv which I believe is a good idea but is not strictly required (helps keep your environment clean and consistent though). And to give credit where credit's due I found this repo from @TheSentry's comment: https://github.com/FriendsOfGalaxy/galaxy-integration-steam/issues/159#issuecomment-1532670017

About the PSN plugin and also probably helpful for you work on the Steam plugin. [Playnite](https://playnite.link/) and its extensions are very well maintained so you could have a look at how they go about things. Here are their Steam and PSN extensions: https://github.com/JosefNemec/PlayniteExtensions/tree/master/source/Libraries/SteamLibrary, https://github.com/XenorPLxx/playnite-library-psn. I'd advise you give them credit if you get any ideas. 

I personally don't have time to help unfortunately. We're slowly pushing GOG Nebula towards the goal of being an org that can maintain these plugins but with no indication from GOG that we could take over @FriendsOfGalaxy or any indication of anything at all, its become extremely slow.

**Reply by ABaumher** [reply]
I was actually on the PlayNite repo, and was trying in C# to get all that workaround stuff working. If i were to implement the new method, it'd basically be a python clone of theirs. 

What irks me is that EA Play (the new Origin replacement) has a sync with Sony that appears pretty much identical to our own, but either their internal browser is modernized or they have a clientid so they can do OAuth properly. It's also closed source, so we can't ~~steal~~ utilize it. On that note, do we know if EA Play going to break the Origin integration?

**Reply by urwrstkn8mare** [reply]
Not sure but I wouldn't be surprised.

**Reply by TheSentry** [reply]
I can't seem to get this process running. How the hell am I supposed to build the plugin? I followed the instructions, but "inv" immediately fails.

As of now, I'm really of no help here. I mean, I see myself as a rather good programmer, but I'm new to Python and I've wasted so many hours on setting up an environment to start developing. I'd love to help, but I can only do that if someone could point me to an actual working How-To on setting up the development environment.

@ABaumher You seem to be making good progress. Can you help me help you? ;)

**Reply by ABaumher** [reply]
@TheSentry Sure. We've been doing our discussions on how the hell we're implementing this whole thing at https://github.com/FriendsOfGalaxy/galaxy-integration-steam/discussions/164. Neither of us there have figured out tasks.py yet, but mostly because we haven't looked. `inv` is shorthand for `invoke`, i don't know if you have that in your virtual environment.

I can post the stuff here/there, or i could try to remember my old discord handle and use that. i stopped using it when they starting requiring phone numbers, but could get back in. 

For the record, i am relatively new to python myself. Coming from a heavy OOP backing (C++, C#, Java), it's... unpleasant.

**Reply by TheSentry** [reply]
@ABaumher The invoke part I could figure out myself. I tried copying the task.py and the other stuff from the buildtools repository, which actualy does stuff, but not what I wanted.

I'm a Java/C# guy myself. Python itself is not so bad, but the development environment is bad. I miss Maven and that says a lot ;)

**Reply by ABaumher** [reply]
It can't be as bad as JGrasp. https://www.jgrasp.org/ First IDE i ever used an i remember it being terrible. Gives me emacs vibes looking back

Have you tried their build tools with tasks.py from https://github.com/thibmo/galaxy-integration-steam/ ? I just managed to merge that into auth_fix (my current branch for all these fixes) and was going to see if invoke worked for me and what it gave me

Edit: I don't remember Ant being terribly fun if i'm being honest. `dotnet publish` isn't too bad

**Reply by ABaumher** [reply]
Update: Using the tasks.py from @thibmo 's fork works - dumps the installation into %localappdata\GOG.com\Galaxy\plugins\installed like it's supposed to. I used powershell, and used Visual Studio to setup the environment, though i think the readme has instructions for doing it manually. 

The act of merging the two repos probably broke a dozen different things, but you can get to testing with a powershell terminal and cloning my repo.
https://github.com/slo/galaxy-integration-steam/tree/AuthFix

lmk if the plugin doesn't work for you

Edit: let me loop in Slo on the dev discussion page. They are busy for the next few days iirc but would probably like to know

**Reply by TheSentry** [reply]
Nice, I'm getting somewhere with this. Still some error messages ("protoc not known"), but I won't have time for that for the next few days.

btw: That repository is actually forked from my FansOfGalaxy project, which I created before knew that the Nebula project existed. So we should straighten that up a bit somehow.

Good work!

**Reply by ABaumher** [reply]
it's possible to merge our code into the one here, though i'd need permission to create a fork here and push it there. alternatively, i can for this repo, merge into that fork, then generate a PR when it's all fixed. I think that's the standard route and probably what I'll do.

Assuming the second approach, the process requires forking off of this repo, then adding my current repo (forked off your stuff) as a remote origin, fetching that branch, then checking it out (to make sure everything looks right), checking out the fork, and merging both branches together, forcing it with a flag (idr, something to allow different histories). I'll get to it eventually.

Protoc is annoying. I could never get it to output the files like they are in our repo or ValvePython's for that matter. Looks like it wasn't part of the repo in any way (it's a standalone exe from Google iirc). But there was a breaking change a few versions back and idk if that's what was giving me fits. I need to see which version ValvePython uses so we can get what we expect

**Reply by ABaumher** [reply]
The protoc binaries can be found here https://github.com/protocolbuffers/protobuf/releases
They're listed in assets below the protobuf files. I think we may need grpc or grpcio, but i haven't the foggiest tbh

**Reply by ABaumher** [reply]
Update: Tasks.py isn't perfect yet; it seems the original used ValvePython's tweaks to make the proto files a little more python-friendly, which aren't in thibmo's version. Which means i need to port a GNU Makefile into a `tasks.py` equivalent. joy. it's just a few `sed` commands, but sed evidently isn't available on Windows without cygwin or a WSL Distro. I'll see what i can do

**Reply by ABaumher** [reply]
Ok, have a version of tasks.py that should pull the protobuf files we need, and convert them. It's on the repo. If they work (the content they generate in the `.py` files are completely different, i have *no idea* if they work identically; hopefully this is just Google refactoring to save file size), then we can generate a full package to send to the Galaxy Plugin library

**Reply by ABaumher** [reply]
It *may* have worked, but i'd rather not try it, at least until this damn thing works and i can start refactoring/modernizing everything. Intead, i found the version of protoc the old devs were using and it generates identical python files. Updated the readmes accordingly.

**Comment by ABaumher** [comment]
Marking this as closed because i have it working. Ty all!
